---
title: "IDEAL epigenetic pilot"
subtitle: "David Martino"
author: |
  Telethon Kids Institute
  University of Western Australia
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    number_sections: true
    toc: yes
    fig_caption: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Project 
Study: IDEAL project  
PI: Ofer Levy  
contact: Ofer Levy <Ofer.Levy@childrens.harvard.edu>   

### Protocol
Cell type: PBMC  
Extraction method: Chemagic low volume blood kit  
Assay: Infinium Human Methylation EPIC array  

### Data Generation  
0.65 - 1.5ug of DNA send to AGRF for bisulphite treatment and hybridization to array  
Contact: Melinda Ziino <melinda.ziino@agrf.org.au>   

# Setup
## Directories
```{r}
datadir <- '/GROUP/RAW_DATA/CAGRF20083570_EPIC_IDEAL'
datadir2 <- '/GROUP/RAW_DATA/CAGRF20083570_EPIC_IDEAL_REPEATS'
outdir <- '/GROUP/ANALYSIS/CAGRF20083570_EPIC_IDEAL'
reportdir <- '/GROUP/ANALYSIS/CAGRF20083570_EPIC_IDEAL/scr'
```

## Libraries 
```{r call library packages, message= FALSE}
library(minfi)
library(RColorBrewer)
library(limma)
library(cluster)
library(ggplot2)
library(reshape2)
library(maxprobes)
library(ggsci)
library(tidyverse)
```

# Importing raw data
```{r prepare sample metadata, message= FALSE}
#read service provider sample annotations
metadata = read.metharray.sheet(datadir, pattern = ".csv")
metadata2 = read.metharray.sheet(datadir2, pattern = ".csv")

```

```{r read in raw data, warning=FALSE}
replace <- which(metadata$Sample_Name %in% metadata2$Sample_Name)
metadata[replace, ] <- metadata2

RGset <- read.metharray.exp(base = NULL, targets = metadata)
```

# Sample level QC
## QC report
```{r fig.width=10, fig.height=8, fig.cap=' ', message= FALSE}
controlStripPlot(RGset, controls="BISULFITE CONVERSION I")
```

```{r fig.width=10, fig.height=8, fig.cap=' '}
controlStripPlot(RGset, controls="BISULFITE CONVERSION II")
```

```{r fig.width=10, fig.height=8, fig.cap=' '}
controlStripPlot(RGset, controls="EXTENSION")
```

```{r fig.width=10, fig.height=8, fig.cap=' '}
controlStripPlot(RGset, controls="HYBRIDIZATION")
```

```{r fig.width=10, fig.height=8, fig.cap=' '}
controlStripPlot(RGset, controls="NON-POLYMORPHIC")
```

```{r fig.width=10, fig.height=8, echo = FALSE, fig.cap=' '}
controlStripPlot(RGset, controls="SPECIFICITY I")
```

```{r fig.width=10, fig.height=8, fig.cap=' '}
controlStripPlot(RGset, controls="SPECIFICITY II")
```

```{r fig.width=10, fig.height=8, fig.cap=' '}
controlStripPlot(RGset, controls="TARGET REMOVAL")
```

## Overall sample quality based on median methylated and unmethylated signals
```{r fig.width=8, fig.height=8, fig.cap = 'Overall sample quality based on methylated and unmethylated signals'}
mSet <- preprocessRaw(RGset)
plotQC(getQC(mSet),badSampleCutoff=10.5)

```

```{r}
#extract qc measures
qc.dat <- getQC(mSet)

#find samples with poor qc
foo <- apply(qc.dat,1,mean)
keep <- which(foo > 10.5)

#remove failed sample
RGset = RGset[, keep]
mSet = mSet[, keep]
metadata = metadata[keep, ]
```


## Checking detection P calls
Average probe p-detection for signal v noise
```{r fig.width=10, fig.height=8, fig.cap = 'Detection p calls'}
detp = detectionP(RGset)

barplot(apply(detp,2,mean), 
        col= 'light blue', 
        xaxt="none", 
        ylim=c(0, 1.5e-2),
        las=2, 
        cex.names=0.8,
        main = 'Average p detection')
abline(h=0.01,col="red")
title(xlab="Arrays",cex=2, line=1)
```

## Raw data distribution
```{r fig.width=8, fig.height=8, fig.cap='Density distribution of raw data coloured by Slide'}
# Define the number of colors you want
nb.cols <- length(unique(metadata$Slide))
mypal <- pal_npg("nrc", alpha = 0.7)(10)

densityPlot(RGset, sampGroups = metadata$Slide, main= "Beta", xlab = "Beta", legend = TRUE, pal = mypal)
```

## Checking for sample mix ups
```{r sex prediction, fig.width=10, fig.height=8, fig.cap = 'MDS analysis of sex', message= FALSE}
#estimate gender from microarray
g.mSet <- mapToGenome(mSet)
sex.pred=getSex(g.mSet)
plot(sex.pred$xMed, sex.pred$yMed, pch= 19, col = ifelse(sex.pred$predictedSex=='M', 'blue','pink'))
text(sex.pred$xMed, sex.pred$yMed, labels = metadata$Sex, pos = 4)
```

# Data normalization
```{r Apply normalisation to raw data, message= FALSE}
#Normalisation = off
gRaw  = preprocessRaw(RGset)
#Normalise stratified quantile (SQN)
gNorm = preprocessQuantile(RGset)
#Normalise Noob
gNoob = preprocessNoob(RGset)
#Normalise SWAN
gSwan = preprocessSWAN(RGset)
```

```{r fig.width=8, fig.height=12, fig.cap='MDS comparison of Normalisation methods'}
par(mfrow = c(2,2))
mdsPlot(getBeta(gRaw),
        sampGroups = metadata$Sample_Group,
        pch=19,legendPos = 'bottomright',
        main = 'Un-normalised')
mdsPlot(getBeta(gNorm),
        sampGroups = metadata$Sample_Group,
        pch=19,legendPos = 'bottomright',
        main = 'SQN')
mdsPlot(getBeta(gNoob),
        sampGroups = metadata$Sample_Group,
        pch=19,legendPos = 'bottomright',
        main = 'Noob')
mdsPlot(getBeta(gSwan),
        sampGroups = metadata$Sample_Group,
        pch=19,legendPos = 'bottomright',
        main = 'Swan')
```

### RLE plots
Relative Log Expression computed for every spot in the array as the difference between the spot intensity and the median intensity for the same feature across all the arrays. Since majority of the spots are not expected to be differentially expressed, the plot shoud show boxplots centered around zero and all of them having the aproximately the same dispersion. An array showing greater dispersion than the other, or being not centered at zero could have quality problems.
```{r fig.width=15, fig.height=12, fig.cap = 'Examining median centered signal pre -post normalization'}
M.raw = getM(gRaw)
mSet.med= apply(M.raw, 1, median)
RLE.raw = M.raw - mSet.med

M.sqn = getM(gNorm)
M.sqn.med = apply(M.sqn, 1, median)
RLE.sqn = M.sqn - M.sqn.med

M.noob = getM(gNoob)
M.noob.med = apply(M.noob, 1, median)
RLE.noob = M.noob - M.noob.med

M.swan = getM(gSwan)
M.swan.med = apply(M.swan, 1, median)
RLE.swan = M.swan - M.swan.med

par(mfrow=c(2,2))
boxplot(RLE.raw, 
        outline=FALSE, 
        ylim=c(-1.5,1.5), 
        ylab="Relative Log Methylation Value",
        col= 'light blue', 
        xaxt="none",
        main='raw data')
title("A",adj=0, cex.main=2)
boxplot(RLE.sqn, 
        outline=FALSE, 
        ylim=c(-1.5,1.5), 
        ylab="Relative Log Methylation Value",
        col= 'light blue', 
        xaxt="none",
        main='sqn')
title("B",adj=0, cex.main=2)
boxplot(RLE.noob, 
        outline=FALSE, 
        ylim=c(-1.5,1.5), 
        ylab="Relative Log Methylation Value",
        col= 'light blue', 
        xaxt="none",
        main='Noob')
title("C",adj=0, cex.main=2)
boxplot(RLE.swan, 
        outline=FALSE, 
        ylim=c(-1.5,1.5), 
        ylab="Relative Log Methylation Value",
        col= 'light blue', 
        xaxt="none",
        main='swan')
title("D",adj=0, cex.main=2)

title(xlab="Arrays",cex=2, line=1)

```

###Density plots

```{r fig.width=10, fig.height=10, fig.cap = 'Examining Beta density distribution of signal pre -post normalization'}
par(mfrow=c(2,2))
densityPlot(getBeta(gRaw), 
            main= "Raw Beta", 
            xlab = "Beta",
            legend=FALSE)
densityPlot(getBeta(gNorm), 
            main= "SQN", 
            xlab = "Beta",
            legend=FALSE)
densityPlot(getBeta(gNoob), 
            main= "NOOB", 
            xlab = "Beta",
            legend=FALSE)
densityPlot(getBeta(gSwan), 
            main= "SWAN", 
            xlab = "Beta",
            legend=FALSE)
```

```{r fig.width=10, fig.height=10, fig.cap = 'Examining M density distribution of signal pre -post normalization'}
par(mfrow=c(2,2))
densityPlot(getM(gRaw), 
            main= "Raw Beta", 
            xlab = "Beta",
            legend=FALSE)
densityPlot(getM(gNorm), 
            main= "SQN", 
            xlab = "Beta",
            legend=FALSE)
densityPlot(getM(gNoob), 
            main= "NOOB", 
            xlab = "Beta",
            legend=FALSE)
densityPlot(getM(gSwan), 
            main= "SWAN", 
            xlab = "Beta",
            legend=FALSE)
```


# Probe-level QC
## Removing poor performing probes
Probes with a detection p-value call of greater than 0.01 in one or more samples were excuded from subsequent analysis
```{r remove poor performing probes}
#choose bad probe cutoff (plim) for filering
#plim= 0.01
plim=1E-16

# ensure probes are in the same order in the gNorm and detP objects
detp <- detp[match(featureNames(gNorm),rownames(detp)),]

#ticker probes before filtering
pre <- length(featureNames(gNorm))

#remove any probes that have failed in one or more samples
keepProbes = rowSums(detp < plim) == ncol(detp)
gNorm = gNorm[keepProbes, ]

#ticker probes after filetering
post <- length(featureNames(gNorm))
```

Probe call rate filter:`r print(plim)`  
In total `r pre-post` probes were removed  
After filtering data set is  
`r ncol(gNorm)` total samples
`r nrow(gNorm)` total probes  

## Remove unwanted SNP and non CG methylation probes
These probes are confounded by genotype or target non-CG methylation
```{r remove SNP probes and non CG methylation probes, message = FALSE}
#ticker number of probes before filtering
pre   <- length(featureNames(gNorm))

gNorm <- dropLociWithSnps(gNorm, snps=c("SBE","CpG"), maf=0)
gNorm <- dropMethylationLoci(gNorm, dropCH = TRUE)

#ticker number of probes after filtering
post  <- length(featureNames(gNorm))
```

In total `r pre-post` probes were removed  
After filtering data set is:  
`r ncol(gNorm)` total samples
`r nrow(gNorm)` total probes

## Remove blacklist probes
The probes have been reported by McCartney DL, et al (2016)  Genom Data 9:22â€“24. to potentially bind  
non-specifically yielding off target effects
```{r remove non-specific probes}
#ticker number of probes before filtering
pre   <- length(featureNames(gNorm))

#drop probes
gNorm = dropXreactiveLoci(gNorm)

#ticker number of probes before filtering
post  <- length(featureNames(gNorm))
```

In total `r pre-post` off target probes were removed  
After filtering data set is:  
`r ncol(gNorm)` total samples
`r nrow(gNorm)` total probes

# Estimation of cell composition
```{r fig.width=12, fig.height=10, fig.cap = 'Reference-based estimation of cell counts'}
library(FlowSorted.Blood.450k)
# ensure that the "Slide" column of the rgSet pheno data is numeric to avoid "estimateCellCounts" error
pData(RGset)$Slide <- as.numeric(pData(RGset)$Slide)

# estimate cell counts
cellCounts <- estimateCellCounts(RGset)

# plot cell type proportions by group
boxplot(cellCounts, xlab = 'Cell Type', ylab = 'Estimated Proportion in Blood', main = 'Estimation of cell counts in blood')
```


# Extract Final Beta and Mvals
```{r extract Beta and Mvals to file}
Beta=getBeta(gNorm)
M=getM(gNorm)
map2genome=getAnnotation(gNorm)

save(RGset, M, Beta, map2genome, metadata, cellCounts, file= file.path(outdir,'savedenvirons.RData'))
```

Final processed data set is:  
`r ncol(Beta)` total samples
`r nrow(Beta)` total probes

## Distribution of processed data
```{r fig.width=10, fig.height=5, fig.cap = 'Processed Beta and M values'}
par(mfrow=c(1,2))
densityPlot(Beta, 
            main= "Beta Values", 
            xlab = "Beta",
            legend=FALSE)
densityPlot(M, 
            main= "M-values", 
            xlab = "M-values",
            legend=FALSE)
```

# R session info
```{r}
sessionInfo()
```
