---
title: "PlinkQC"
author: "David Martino  \nTelethon Kids Institute  \nUniversity of Western Australia\n"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
  html_document:
    toc: yes
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
```

**Project**
*Study: IDEAL project*  
*PI: Ofer Levy*  
*contact: Ofer Levy <Ofer.Levy@childrens.harvard.edu>*   

**Protocol**
*Cell type: PBMC*  
*Extraction method: Chemagic low volume blood kit*  
*Assay: Infinium Global Screening array + MD*  

**Data Generation**  
*0.65 - 1.5ug of DNA send to AGRF for bisulphite treatment and hybridization to array*  
*Contact: Melinda Ziino <melinda.ziino@agrf.org.au>*  
 

# 0.0 Setup

## Libraries 
```{r call library packages, message= FALSE}
library(plinkQC)
library(tidyverse)
library(UpSetR)
```

The PLINKQC package was designed to be used with plink v1.9

Workflow begins by computing joint PCA of study population and reference datasets from the 1000 Genomes project. 

# 1.0 Create binary plink file set and filter both data sets for non A-T or G-C SNPs

using the PLINK command to convert the .ped and .map files to binary format using the --make-bed option. The output binary files will be created in the same directory as the input files, with the same name as the input file, but with different extensions such as .bed, .bim, .fam.

```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir='/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
name='ILGSAMD-17768'

plink --file  $qcdir/$name \
      --make-bed \
      --out $qcdir/$name
```


A suitable reference dataset should be downloaded and if necessary, re-formated into PLINK format. see https://meyer-lab-cshl.github.io/plinkQC/articles/Genomes1000.html for details on obtaining the reference data set.

This script is using the PLINK tool to filter out SNPs that have a certain genotype (AT, TA, GC, or CG) from two sets of data (stored in the variables $qcdir/$name and $refdir/$refname). The script is applied to a binary file set (bed/bim) and uses awk to extract the SNP IDs that match the specified genotype from the .bim files of the two datasets. These files are then used to exclude the corresponding SNPs from the datasets using PLINK's --exclude option. The resulting datasets with the excluded SNPs are saved in new files ($qcdir/$name.no_ac_gt_snps and $qcdir/$refname.no_ac_gt_snps). Additionally, the script also moves the log files generated by PLINK to a separate directory ($qcdir/plink_log).


```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir='/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
refdir='/mnt/scratch/reference'
name='ILGSAMD-17768'
refname='all_phase3'

mkdir $qcdir/plink_log

awk 'BEGIN {OFS="\t"}  ($5$6 == "GC" || $5$6 == "CG" || $5$6 == "AT" || $5$6 == "TA")  {print $2}' $qcdir/$name.bim > $qcdir/$name.ac_gt_snps
awk 'BEGIN {OFS="\t"}  ($5$6 == "GC" || $5$6 == "CG" || $5$6 == "AT" || $5$6 == "TA")  {print $2}' $refdir/$refname.bim > $qcdir/$refname.ac_gt_snps

plink --bfile  $refdir/$refname \
      --exclude $qcdir/$refname.ac_gt_snps \
      --allow-extra-chr \
      --make-bed \
      --out $qcdir/$refname.no_ac_gt_snps
mv  $qcdir/$refname.no_ac_gt_snps.log $qcdir/plink_log/$refname.no_ac_gt_snps.log

plink --bfile  $qcdir/$name \
      --exclude $qcdir/$name.ac_gt_snps \
      --make-bed \
      --out $qcdir/$name.no_ac_gt_snps
mv  $qcdir/$name.no_ac_gt_snps.log $qcdir/plink_log/$name.no_ac_gt_snps.log
```

# 2.0 Prune study data

We will conduct principle component analysis on genetic variants that are pruned for variants in linkage disequilibrium (LD) with an 𝑟2>0.2 in a 50kb window. The LD-pruned dataset is generated below, using plink –indep-pairwise to compute the LD-variants; additionally exclude range is used to remove genomic ranges of known high-LD structure. 


## find location of high ld regions file

```{r}
file.path(find.package('plinkQC'),'extdata','high-LD-regions-hg19-GRCh37.txt')
```

now that we know that file path, we use it as the highld variable below

```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir='/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
refdir='/mnt/scratch/reference'
name='ILGSAMD-17768'
refname='all_phase3'
highld="/home/dmartino/R/x86_64-pc-linux-gnu-library/4.2/plinkQC/extdata/high-LD-regions-hg19-GRCh37.txt"

plink --bfile  $qcdir/$name.no_ac_gt_snps \
      --exclude range $highld \
      --indep-pairwise 50 5 0.2 \
      --out $qcdir/$name.no_ac_gt_snps
mv  $qcdir/$name.prune.log $qcdir/plink_log/$name.prune.log

plink --bfile  $qcdir/$name.no_ac_gt_snps \
      --extract $qcdir/$name.no_ac_gt_snps.prune.in \
      --make-bed \
      --out $qcdir/$name.pruned
mv  $qcdir/$name.pruned.log $qcdir/plink_log/$name.pruned.log

```

# 3.0 Filter reference data for the same SNP set as in study

extract a subset of SNPs from a binary PLINK file specified by the --bfile flag with the file name "$refdir/$refname" using a list of SNPs specified by the --extract flag with the file name "$qcdir/$name.no_ac_gt_snps.prune.in". The command then creates a new binary PLINK file with the subset of SNPs using the --make-bed option and saves the output to a file with the name "$qcdir/$refname.pruned".


```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir='/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
refdir='/mnt/scratch/reference'
name='ILGSAMD-17768'
refname='all_phase3'

plink --bfile  $refdir/$refname \
      --extract $qcdir/$name.no_ac_gt_snps.prune.in \
      --make-bed \
      --allow-extra-chr \
      --out $qcdir/$refname.pruned
mv  $qcdir/$refname.pruned.log $qcdir/plink_log/$refname.pruned.log

```

# 4.0 Checking and correcting any chromosome / position mismatch

## mismatched chroms

The first command is using awk, a scripting language for text processing, to compare the rs numbers in two files: "$qcdir/$name.pruned.bim" and "$qcdir/$refname.pruned.bim" and creating a new file "$qcdir/$refname.toUpdateChr" with the rs numbers that have different chromosome numbers in these two files. It also filters out any lines that contain "XY" using sed and the -n option.

The second command is using PLINK to update the chromosome numbers of the SNPs in the file "$qcdir/$refname.pruned" based on the information in the file "$qcdir/$refname.toUpdateChr" using the --update-chr flag, where the first and second columns of the update file are used to specify the old and new chromosome numbers, respectively. The command then creates a new binary PLINK file with the updated chromosome numbers using the --make-bed option and saves the output to a file with the name "$qcdir/$refname.updateChr"

```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir='/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
refdir='/mnt/scratch/reference'
name='ILGSAMD-17768'
refname='all_phase3'


awk 'BEGIN {OFS="\t"} FNR==NR {a[$2]=$1; next} \
    ($2 in a && a[$2] != $1)  {print a[$2],$2}' \
    $qcdir/$name.pruned.bim $qcdir/$refname.pruned.bim | \
    sed -n '/^[XY]/!p' > $qcdir/$refname.toUpdateChr

plink --bfile $qcdir/$refname.pruned \
      --update-chr $qcdir/$refname.toUpdateChr 1 2 \
      --make-bed \
      --out $qcdir/$refname.updateChr
mv $qcdir/$refname.updateChr.log $qcdir/plink_log/$refname.updateChr.log

```

## Mismatched positions

This command is using awk, a scripting language for text processing, to compare the rs numbers in two files: "$qcdir/$name.pruned.bim" and "$qcdir/$refname.pruned.bim" and creating a new file "$qcdir/$refname.toUpdatePos" with the rs numbers that have different position in these two files.

The awk script is using the following logic:

It sets the output field separator to tab character ("\t")
In the first file, it creates an associative array "a" with the second column ($2) as the key and the fourth column ($4) as the value
In the second file, for each line, it checks if the second column ($2) is present in the array "a", and if the value associated with that key is different from the fourth column ($4) of the current line, it prints the value associated with that key and the second column of the current line


```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir='/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
refdir='/mnt/scratch/reference'
name='ILGSAMD-17768'
refname='all_phase3'

awk 'BEGIN {OFS="\t"} FNR==NR {a[$2]=$4; next} \
    ($2 in a && a[$2] != $4)  {print a[$2],$2}' \
    $qcdir/$name.pruned.bim $qcdir/$refname.pruned.bim > \
    $qcdir/${refname}.toUpdatePos
```

##  possible allele flips

This script is using AWK to find SNPs that need to be flipped in the dataset. It is comparing two files, $qcdir/$name.pruned.bim and $qcdir/$refname.pruned.bim, and looking for SNPs with the same identifier ($1$2$4) in both files. If the genotype of the SNP in the first file ($5$6) is different from the genotype in the second file ($5$6) and the reverse of the genotype in the second file ($6$5), it is printing the SNP name ($2) to the output file $qdir/$refname.toFlip.
It's using the FNR variable to compare first file against second file.

```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir='/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
refdir='/mnt/scratch/reference'
name='ILGSAMD-17768'
refname='all_phase3'

awk 'BEGIN {OFS="\t"} FNR==NR {a[$1$2$4]=$5$6; next} \
    ($1$2$4 in a && a[$1$2$4] != $5$6 && a[$1$2$4] != $6$5)  {print $2}' \
    $qcdir/$name.pruned.bim $qcdir/$refname.pruned.bim > \
    $qcdir/$refname.toFlip

```

## Upate positions and flip alleles

This script is using PLINK to update the map positions and flip the alleles for the SNPs in the dataset $qcdir/$refname.updateChr. The --update-map flag is used to update the map positions of SNPs in the file $qcdir/$refname.toUpdatePos, where the first column is the SNP name and the second and third columns are the new chromosome and position, respectively. The --flip flag is used to flip the alleles of SNPs in the file $qcdir/$refname.toFlip. The --make-bed flag is used to create a new binary PLINK file. The output of this command is saved in the file $qdir/$refname.flipped and the log file is moved to the directory $qcdir/plink_log.
This command is updating the position of SNPs and flipping the alleles of SNPs as per the updated map positions and flipped alleles stored in respective files.

```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir='/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
refdir='/mnt/scratch/reference'
name='ILGSAMD-17768'
refname='all_phase3'

plink --bfile $qcdir/$refname.updateChr \
      --update-map $qcdir/$refname.toUpdatePos 1 2 \
      --flip $qcdir/$refname.toFlip \
      --make-bed \
      --out $qcdir/$refname.flipped
mv $qcdir/$refname.flipped.log $qcdir/plink_log/$refname.flipped.log
```

## Remove mismatches

Any alleles that do not match after allele flipping, are identified and removed from the reference dataset.

```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir='/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
refdir='/mnt/scratch/reference'
name='ILGSAMD-17768'
refname='all_phase3'

awk 'BEGIN {OFS="\t"} FNR==NR {a[$1$2$4]=$5$6; next} \
    ($1$2$4 in a && a[$1$2$4] != $5$6 && a[$1$2$4] != $6$5) {print $2}' \
    $qcdir/$name.pruned.bim $qcdir/$refname.flipped.bim > \
    $qcdir/$refname.mismatch

plink --bfile $qcdir/$refname.flipped \
      --exclude $qcdir/$refname.mismatch \
      --make-bed \
      --out $qcdir/$refname.clean
mv $qcdir/$refname.clean.log $qcdir/plink_log/$refname.clean.log

```

# Merge reference data and study pop data

This script is using PLINK to merge the dataset $qcdir/$name.pruned with the dataset $qcdir/$refname.clean.
It uses the --bfile flag to specify the input dataset $qcdir/$name.pruned and the --bmerge flag to specify the files for the second dataset: $qcdir/$refname.clean.bed, $qcdir/$refname.clean.bim, and $qcdir/$refname.clean.fam. The --make-bed flag is used to create a new binary PLINK file.

```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir='/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
refdir='/mnt/scratch/reference'
name='ILGSAMD-17768'
refname='all_phase3'

plink --bfile $qcdir/$name.pruned  \
      --bmerge $qcdir/$refname.clean.bed $qcdir/$refname.clean.bim \
         $qcdir/$refname.clean.fam  \
      --make-bed \
      --out $qcdir/$name.merge.$refname
mv $qcdir/$name.merge.$refname.log $qcdir/plink_log

```

# PCA on merged data

We can now run principal component analysis on the combined dataset using plink –pca which returns a .eigenvec file with the family and individual ID in columns 1 and 2, followed by the first 20 principal components.

```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir='/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
refdir='/mnt/scratch/reference'
name='ILGSAMD-17768'
refname='all_phase3'

plink --bfile $qcdir/$name.merge.$refname \
      --pca \
      --out $qcdir/$name.$refname
mv $qcdir/$name.$reference.log $qcdir/plink_log

```


```{r fig.width=10, fig.height=8}

indir <- '/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
name <- 'ILGSAMD-17768'
refname <- 'all_phase3'
prefixMergedDataset <- paste(name, ".", refname, sep="")

#get the refsamplesfile
refSamplesFile =  paste(system.file("extdata", package="plinkQC"),"/Genomes1000_ID2Pop.txt", sep="")

#corresponding colors file
refColors = read_csv(file.path(indir,'refSampleColors.csv')) %>%
  as.data.frame()

#plotting function
exclude_ancestry <- evaluate_check_ancestry(indir=indir, name=name, europeanTh = 3,
                            prefixMergedDataset=prefixMergedDataset,
                            refSamplesFile=refSamplesFile, 
                            refColors=refColors,
                            studyColor = 'red',
                            legend_text_size = 15,
                            legend_title_size = 20,
                            axis_text_size = 20,
                            axis_title_size = 20,
                            interactive=TRUE)

```


# 5.0 Per-individual QC

At this point in the script we use qcdir as the directory for processed data

```{r}
indir <- '/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154'
qcdir <- '/mnt/fws_hpc_cepi/ANALYSIS/CAGRF20083570_GSA_MD_IDEAL'
name <- 'ILGSAMD-17768'
refname <- 'all_phase3'
prefixMergedDataset <- paste(name, ".", refname, sep="")
path2plink <- "/home/dmartino/bin/plink_linux_x86_64_20221210/plink"

#get the refsamplesfile
refSamplesFile =  paste(system.file("extdata", package="plinkQC"),"/Genomes1000_ID2Pop.txt", sep="")

#corresponding colors file
refColors = read_csv(file.path(indir,'refSampleColors.csv')) %>%
  as.data.frame()

#get AGRF sample sheet
#samp.sheet <- read.csv('/mnt/fws_hpc_cepi/RAW_DATA/Project_TAHS/CAGRF20083570_GSA_MD_TAHS/CAGRF20083570_GSA-MD_TAHS_SampleSheet.csv', skip=8)

#samp.sheet$Sex = ifelse(samp.sheet$Sex=='male', 0,1)
```

## Check discordant sex
```{r fig.height = 5, fig.width = 8}
fail_sex <- check_sex(indir=indir, qcdir=indir, name='ILGSAMD-17768',
                      #externalSex = samp.sheet,
                      #externalFemale = 1,
                      #externalMale = 0,
                      #externalSexSex = "Sex",
                      #externalSexID = "Sample_ID",
                      interactive=TRUE,
                      legend_text_size = 10,
                      verbose=TRUE, 
                      path2plink=path2plink)
```

>**Following individuals are discordant for sex**

```{r}
print(fail_sex$fail_sex)
```

## Check missing genotypes / heterozygosity rate
```{r fig.height = 5, fig.width = 8}
fail_het_imiss <- check_het_and_miss(indir=indir, qcdir=indir, name='ILGSAMD-17768',
                                     legend_text_size = 10, run.check_het_and_miss=TRUE,
                                     interactive=TRUE, 
                                     path2plink=path2plink)
```

>**Following individuals failed heterozygosity test**

```{r}
print(fail_het_imiss$fail_het)
```

>**Following individuals outlying missing genotypes**

```{r}
print(fail_het_imiss$fail_imiss)
```

## Check relatedness
Need to first generate a genome file. The option --genome generates genome-wide identity-by-descent (IBD) sharing probabilities. It is used for estimating pairwise relatedness between all pairs of samples in a dataset, as well as for identifying and removing closely related samples.

```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir="/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154"
outdir="/mnt/fws_hpc_cepi/RAW_DATA/CAGRF20083570_GSAMD_IDEAL/PLINK_010422_1154"
name='ILGSAMD-17768'

plink --file $qcdir/$name \
      --genome \
      --out $outdir/$name
      
```


```{r fig.height=5, fig.width=8}
exclude_relatedness <- check_relatedness(indir=indir, qcdir=indir, name='ILGSAMD-17768',
                                         legend_text_size = 10,
                                         axis_text_size = 10,
                                         axis_title_size = 10,
                                         interactive=TRUE,
                                         showPlinkOutput = FALSE,
                                         path2plink=path2plink)
```

>**Following are highly related samples**

```{r}
print(exclude_relatedness$fail_highIBD)
```

## Overview of sample QC fail

```{r}
#first chain together the outputs
fail_list <- list(missing_genotype = fail_het_imiss$fail_imiss[,2],
            high_IBD = exclude_relatedness$failIDs[,2],
            outlying_heterozygosity=fail_het_imiss$fail_het[,2],
            mismatched_sex=fail_sex$fail_sex[,2],
            mismatched_ancestry=exclude_ancestry$fail_ancestry[,2])

print(fail_list)
```

```{r message=FALSE}
upset(fromList(fail_list), 
      order.by = "freq", 
      text.scale = c(1.3, 1.3, 1, 1, 2, 0.75),
      sets.bar.color = "#56B4E9")
```

```{r fig.width=10, fig.height=10}

fail_individuals <- perIndividualQC(indir=indir, qcdir=indir, name="ILGSAMD-17768", path2plink=path2plink,
                            do.evaluate_check_relatedness = TRUE,
                            do.evaluate_check_sex = TRUE,
                            do.evaluate_check_het_and_miss = TRUE,
                            dont.check_ancestry = TRUE,
                            interactive=TRUE, verbose=FALSE, showPlinkOutput = FALSE)

```

# 6.0 Per-marker quality control
## Markers with excessive missing rate


```{r  fig.height=5, fig.width=8}
fail_snpmissing <- check_snp_missingness(indir=indir, qcdir=indir, name="ILGSAMD-17768",
                                         interactive=TRUE,
                                         path2plink=path2plink,
                                         legend_text_size = 10,
                                         axis_text_size = 10,
                                         axis_title_size = 10,
                                         showPlinkOutput=FALSE)
```

>**Overall missing SNP rate was 3.2%**  

## Markers with deviation from HWE
```{r  fig.height=5, fig.width=8}
fail_hwe <- check_hwe(indir=indir, qcdir=indir, name="ILGSAMD-17768", 
                      interactive=TRUE,
                      path2plink=path2plink,
                      legend_text_size = 10,
                      axis_text_size = 10,
                      axis_title_size = 10,
                      showPlinkOutput=TRUE)
```

## Markers with low MAF

filtering out ultra-low variants (< 0.05 MAF) which dont perform well in imputation according to Verlouw et al 2021 Eu J hum Gen
```{r}
fail_maf <- check_maf(indir=indir, qcdir=indir, name=name, interactive=TRUE, mafTh = 0.005,
                      path2plink=path2plink, showPlinkOutput=TRUE)
```


## Overview of marker QC fail

```{r}
#first chain together the outputs
fail_list <- list(missing_calls = fail_snpmissing$fail_missingness[,2],
            fail_hew = fail_hwe$fail_hwe [,2], fail_maf = fail_maf$fail_maf[,2])

upset(fromList(fail_list), 
      order.by = "freq", 
      text.scale = c(1.3, 1.3, 1, 1, 2, 0.75),
      sets.bar.color = "#56B4E9")
```


```{r fig.width=10, fig.height=10}

fail_markers <- perMarkerQC(indir=indir, qcdir=indir, name=name,
                            path2plink=path2plink,
                            mafTh = 0.005,
                            verbose=TRUE, interactive=TRUE,
                            showPlinkOutput=FALSE)

```

# 7.0 Final QC'd data set
```{r}
Ids  <- cleanData(indir=indir, qcdir=indir, name=name, path2plink='/home/dmartino/bin/plink_linux_x86_64_20221210/plink',
                  filterSex = TRUE,
                  filterHeterozygosity = TRUE,
                  filterSampleMissingness = TRUE,
                  filterAncestry = FALSE,
                  filterRelated = FALSE,
                  filterSNPMissingness = TRUE,
                  mafTh = 0.005,
                  verbose=TRUE, showPlinkOutput=TRUE)
```

## Move cleaned data to analysis folder

```{r}
# Set the source and target directories
source_dir <- indir
target_dir <- qcdir

# Set the prefix of the files to be moved
prefix <- "ILGSAMD-17768.clean"

# List all files in the source directory that match the prefix
files <- list.files(source_dir, pattern = paste0("^", prefix))

# Use a for loop to copy each file to the target directory
for (file_name in files) {
  # Construct the full paths of the source and target files
  source_file <- file.path(source_dir, file_name)
  target_file <- file.path(target_dir, file_name)
  
  # Use file.copy() to copy the file to the target directory
  file.copy(source_file, target_file)
  
  # Use file.remove() to remove the original file from the source directory
  file.remove(source_file)
}


```


```{r}
print(Ids)
```
## Report final genotypes and individuals

```{bash}
export PATH=$PATH:/home/dmartino/bin/plink_linux_x86_64_20221210

qcdir="/mnt/fws_hpc_cepi/ANALYSIS/CAGRF20083570_GSA_MD_IDEAL"
name='ILGSAMD-17768.clean'

plink --bfile  $qcdir/$name \
      --check-sex
  
```



# R Session Info
```{r}
sessionInfo()
```
